src/net_processing.cpp

```diff
@@ -4023,10 +4023,14 @@ bool PeerManagerImpl::MaybeDiscourageAndDisconnect(CNode& pnode)
         return false;
     }
 
-    if (pnode.addr.IsLocal()) {
-        // We disconnect local peers for bad behavior but don't discourage (since that would discourage
-        // all peers on the same local address)
-        LogPrintf("Warning: disconnecting but not discouraging local peer %d!\n", peer_id);
+    if (pnode.addr.IsLocal() || pnode.m_inbound_onion) {
+        // We disconnect local or onion peers for bad behavior but don't discourage
+        // (since that would discourage all peers on the same address)
+        if (pnode.m_inbound_onion) {
```

Comment:
micro-nit, could save a conditional and a level of nesting by testing for `IsLocal()` and then for `m_inbound_onion`